dataflowengineoss/src/main/scala/io/shiftleft/dataflowengineoss/passes/reachingdef/ReachingDefPass.scala

```diff
@@ -91,32 +88,88 @@ class ReachingDefPass(cpg: Cpg) extends ParallelCpgPass[nodes.Method](cpg) {
-
-    def getGensOfCall(call: nodes.Call): Set[nodes.StoredNode] = {
-      val methodParamOutsOrder = callToMethodParamOut(call)
+    def defsMadeByCall(call: nodes.Call): Set[nodes.StoredNode] = {
+      val indicesOfDefinedOutputParams = NoResolve
+        .getCalledMethods(call)
+        .flatMap(method => method.parameter.asOutput)
         .filter(methPO => methPO._propagateIn().hasNext)
         .map(_.asInstanceOf[nodes.HasOrder].order.toInt)
-      filterArgumentIndex(call._argumentOut().asScala.toList, methodParamOutsOrder).toSet
+      call
+        ._argumentOut()
+        .asScala
+        .toList
+        .filter(node =>
+          indicesOfDefinedOutputParams.exists(_ == node.asInstanceOf[nodes.HasArgumentIndex].argumentIndex.toInt))
+        .toSet ++ {
+        if (methodForCall(call)
+              .map(method => method.methodReturn)
+              .exists(methodReturn => methodReturn._propagateIn().hasNext) ||
+            !hasAnnotation(call)) {
+          Set(call)
+        } else {
+          Set()
+        }
+      }
     }
 
-    val gensForParams = method.start.parameter.l.map { param =>
+    val defsForParams = method.start.parameter.l.map { param =>
       param -> Set(param.asInstanceOf[nodes.StoredNode])
-    }.toMap
+    }
 
-    val gensForCalls = method.start.call.map { call =>
-      call -> getGensOfCall(call)
+    val defsForCalls = method.start.call.l.map { call =>
+      call -> defsMadeByCall(call)
     }
 
-    (gensForParams ++ gensForCalls).toMap
+    (defsForParams ++ defsForCalls).toMap
   }
 
   def initKill(method: nodes.Method,
                gen: Map[nodes.StoredNode, Set[nodes.StoredNode]]): Map[nodes.StoredNode, Set[nodes.StoredNode]] = {
+
+    def allOtherInstancesOf(node: nodes.StoredNode): Set[nodes.StoredNode] = {
+      declaration(node)
+        .flatMap(instances(_).headOption)
+        .filter(_.id2 != node.id2)
+        .toSet
+    }
+
     method.start.call.map { call =>
-      call -> gen(call).map(v => killsVertices(v)).fold(Set())((v1, v2) => v1.union(v2))
+      call -> gen(call).map(v => allOtherInstancesOf(v)).fold(Set())((v1, v2) => v1.union(v2))
     }.toMap
   }
 
+  def instances(decl: nodes.StoredNode): List[nodes.StoredNode] = {
+    decl._refIn().asScala.toList
+  }
+
+  def uses(node: nodes.StoredNode, gen: Map[nodes.StoredNode, Set[nodes.StoredNode]]): Set[nodes.StoredNode] = {
+    node match {
+      case ret: nodes.Return =>
+        ret.astChildren.map(_.asInstanceOf[nodes.StoredNode]).toSet()
+      case call: nodes.Call =>
+        val parameters = NoResolve.getCalledMethods(call).headOption.map(_.parameter.l).getOrElse(List())
+        node
+          ._argumentOut()
+          .asScala
+          .filter(
+            arg =>
+              parameters
+                .filter(_.order == arg.asInstanceOf[HasOrder].order)
+                .flatMap(_._propagateOut().asScala.toList)
+                .nonEmpty || !gen(node).contains(arg))
```

Comment:
nbd, but could make this into an _and_ to make it easier to parse

Reply:
Yeah, I'm bringing in some readability improvements + more tests and comments today.